### PR TITLE
Feat/tflite compatibility

### PR DESCRIPTION
**Pull request by kenders2000**
Hi there!

Thanks for this great package: I have been using it for a little while and found it really useful.

I have added a TFLITE compatible option to STFT. I have included some unit tests that carry out the tflite conversion of the model, infer and compare with the predictions (tensorflow) of a non tflite compatible model. 

The idea behind this is that you still produce training data / train a model using `STFT(tflite_compatible=False)` (which is faster than the tflite compatible model) and then for model deployment  use `STFT(tflite_compatible=True)` and just transfer the weights to the new model.

`tflite_compatible_stft.py` contains equivalent functions to; tf.signal.stft (`stft_tflite`) (which uses `_rdft_matrix` to calculate the DFT instead of `tf.signal.rfft`); and tf.signal.frame (`fixed_frame`). 

The tflite version of `STFT` returns an extra dimension  (batch, ch, time, freq, re/imag), as tflite does not have complex types, I updated the `Magnitude` and `Phase` layers so that they will detect this extra dimension and process the tensors appropriately. The tflite model also requires a fixed batch size.

There is also no tflite op for `atan`, (that I could find) so I included an approximate version so that phase can at least be approximately calculated. The phase layer can be initialised as follows `Phase(approx_atan_accuracy=500)` approx_atan_accuracy represents the number of iterations used to calculate atan, so there is trade off between speed and accuracy, that can be made based on the application requirements.

I have not looked at making the `InverseSTFT` Tflite compatible, if that is something that would be useful to people I would be happy to look at this sometime?

Let me know if there is anything else that you need to assess this.

Cheers!

Paul

### CODE REVIEW

**Comment by keunwoochoi** [comment]
Hi, sorry for a late reply. I'll review this in a week.

**Comment by Path-A** [comment]
Not sure if this will eventually change anything, but [Tensorflow's 2.5 release markdown](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) currently has this line:

- ```Add RFFT2D as builtin op. (RFFT2D also supports RFFTD.) Currently only supports float32 input.```

File: kapre/tflite_compatible_stft.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
+# -*- coding: utf-8 -*-
```
Comment (keunwoochoi) [inline kapre/tflite_compatible_stft.py:1]: - maybe this is not necessary? 
- comment looks great. let's wrap it with `""" """`.

**Review by keunwoochoi — COMMENTED** [review]
The implementation and tests seem pretty much complete. I'd like to discuss the API changes though.

`STFT` layer would behave the same by default, so it looks all good to me. But `Magnitude` and `Phase` aren't, and I'm not sure how much I can assume that the input of those layers would be always 4-dim batch. Probably that's the case for most of the times, but I just can't be sure. 

How about add a `tflite_compatible.py` and add implementations of all of these (STFT, Mag, etc)? That would be definitely safer, and I expect the implementation would be not too complicate with inheritance. Of course it would be some work on the top of things that are already done though.

Any thought?

File: kapre/tflite_compatible_stft.py (hunk 1 above)
Comment (kenders2000) [reply kapre/tflite_compatible_stft.py:1]: Thanks for getting back to me! 

I had two ideas for the API, one using a `tflilite_compatible` bool argument for Magnitude and phase, but I thought that would get messy and complicated, hence I opted for the dimension checking. The STFT layer as it stands enforces the dimensionality (the transpose ln 193 - 198), which is why I thought it would be ok. However, I like your suggestion regarding inheritance and would be happy to put some more work into this, not a problem!

Your suggestion has the added benefit of when tensorflow adds these ops to tflite, as @Path-A pointed out, we can simply just delete the module!

I will have a go at your suggestion.

Regards

Paul

Comment (keunwoochoi) [reply kapre/tflite_compatible_stft.py:1]: perfect, thanks!

Comment (kenders2000) [reply kapre/tflite_compatible_stft.py]: I've moved all tflite layers to a separate module `time_frequency_tflite.py` all the layer inherits from corresponding layers in `time_frequency`.

the only thing that remains in `time_frequency` is the option to compute the approximate phase in the `Phase` layer. (so that you can generate training data with the approximation)

Let me know what you think

Cheers!

Paul

**Review by keunwoochoi — APPROVED** [review]
Looks good to me for the moment. I'd need to merge it to a branch to see how docstrings are rendered so I'm merging it to `keras-0.3.5.` which I'll need to create anyway. Then I'll have another look on local and merge that to master.

Thanks for the work!

**Comment by keunwoochoi** [comment]
oh, just to make sure.. it's ready to be merged on your side, right?

**Comment by kenders2000** [comment]
> oh, just to make sure.. it's ready to be merged on your side, right?

Yes, I've run the branch to generate a few models, and it is working. Ready from my side, let me know if there is anything further I help with. I will look at the inverse transform next.

Cheers!

Paul

**Comment by keunwoochoi** [comment]
Awesome!
